language/move-symbol-pool/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A global, uniqued cache of strings that is never purged.
+//!
+//! This module provides storage for strings that are meant to remain in use for
+//! the entire running duration of a program. Strings that are stored in this
+//! global, static cache are never evicted, and so the memory consumed by them
+//! can only ever grow.
+//!
+//! The strings can be accessed via the [`Symbol`] type, which acts like a
+//! pointer to the underlying cached string value.
+//!
+//! NOTE: If you're looking for a `#[forbid(unsafe_code)]` attribute here, you
+//! won't find one: symbol-pool (and its inspiration, servo/string-cache) uses
+//! `unsafe` Rust in order to store and dereference `Symbol` pointers to
+//! strings.
+//!
+//! [`Symbol`]: crate::move_symbol_pool::Symbol
+
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::{cmp::Ordering, collections::HashSet, fmt, num::NonZeroU64, sync::Mutex};
+
+type SymbolData = Box<str>;
+static SYMBOL_POOL: Lazy<Mutex<HashSet<SymbolData>>> = Lazy::new(|| Mutex::new(HashSet::new()));
```

Comment:
Thoughts on a `RwLock` v. `Mutex`?

I initially thought a `RwLock` would be better but the lock isn't actually used for reading so maybe the overhead is useless.

Reply:
The latest implementation reads and writes, and so the `RwLock` is definitely an improvement -- thanks!